data_prep/dp/generate_climos.py

```diff
@@ -0,0 +1,357 @@
+
+        :param cdo_op: (function) applies the desired CDO operation. Invoked only with kw args; CDO operation-specific
+            args should be curried by cdo_op. Input and output file args are supplied as kw args here
+        :param output_file_path: (str) specifies file path for output file; it will be created by this function.
+        """
+        try:
+            logger.info('Output file: {}'.format(output_file_path))
+            if not os.path.exists(os.path.dirname(output_file_path)):
+                os.makedirs(os.path.dirname(output_file_path))
+            cdo_op(input=climo_means, output=output_file_path)
+        except Exception as e:
+            logger.warn('Failed to create climatology file. {}: {}'.format(e.__class__.__name__, e))
+        else:
+            output_file_paths.append(output_file_path)
+
+    if split_vars and len(input_file.dependent_varnames) > 1:
+        # Split climo means file into separate files, one per variable
+        logger.info('Splitting into single-variable files')
+        for variable in input_file.dependent_varnames:
+            output_file_path = climo_output_filepath(outdir, input_file, t_start, t_end, variable=variable)
+            create_output_file(lambda **io: cdo.select('name={}'.format(variable), **io), output_file_path)
+    else:
+        # Don't split; copy the temporary file to the final output filename
+        output_file_path = climo_output_filepath(outdir, input_file, t_start, t_end)
+        create_output_file(lambda **io: cdo.copy(**io), output_file_path)
+
+    # TODO: fix <variable_name>:cell_methods attribute to represent climatological aggregation
+    # TODO: Does the above TODO make any sense? Each variable has had up to 3 different aggregations applied
+    # to it, and there is no standard cell_method string that expresses more than one.
+
+    return output_file_paths
+
+
+def generate_climo_time_var(t_start, t_end, types={'monthly', 'seasonal', 'annual'}):
+    """Generate information needed to update the climatological time variable.
+
+    :param t_start: (datetime.datetime) start date of period over which climatological means are formed
+    :param t_end: (datetime.datetime) end date of period over which climatological means are formed
+    :param types: (set) specifies what means have been generted, hence which time values to generate
+    :returns: (tuple) times, climo_bounds
+        times: (list) datetime for *center* of each climatological mean period; see CF standard
+        climo_bounds: (list) bounds (start and end date) of each climatological mean period
+
+    ASSUMPTION: Time values are in the following order within the time dimension variable.
+        monthly: 12 months in their usual order
+        seasonal: 4 seasons: DJF, MAM, JJA, SON
+        annual: 1 value
+    """
+
+    # Year of all time values is middle year of period
+    year = (t_start + (t_end - t_start)/2).year + 1
+
+    times = []
+    climo_bounds = []
+
```

Comment:
Do the computations of climatological times for month and season really have to be this complicated? Examples in the [CF Metadata Standard](http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#climatological-statistics) consistently use the 15th day of each month for multi-year monthly averages and the 16th day of the mid-season months (Jan, Apr, July, Oct) for multi-year seasonal averages. Any reason not to use these?